File: Swift2/Sources/Swift2/LineValueSequencer.swift
public class CLineValueSequencer : CValueSequencer
{
	public init(nValue: Int) {
		super.init(nInnerSequencers: 3, nValues: 4, nValue: nValue)
		
		m_pvsInner = [CBitValueSequencer(nValue: 0) as CValueSequencer,CBitValueSequencer(nValue: 0) as CValueSequencer,
			CBitValueSequencer(nValue: 0) as CValueSequencer]
		//m_pvsInner?[2] = CBitValueSequencer(nValue: 0)
 		//m_pvsInner?[1] = CBitValueSequencer(nValue: 0)
 		//m_pvsInner?[0] = CBitValueSequencer(nValue: 0)
		Bit(nIndex: 2).SetParent(pvsParent: self)
 		Bit(nIndex: 1).SetParent(pvsParent: self)
 		Bit(nIndex: 0).SetParent(pvsParent: self)
 		//m_nSequences = Sequences.nLineSequences
 		//m_nRatios = Sequences.nLineRatios
		
	}

	public func Bit(nIndex: Int) -> CBitValueSequencer {
 		return m_pvsInner![nIndex] as! CBitValueSequencer
 	}

	@discardableResult
	public override func Inverse() -> CValueSequencer {
 		switch (Value)
 		{
 	 		case 0:
 	 			Value = 1
 	 			//break
 	 		case 1:
 	 			Value = 0
 	 			break
 	 		case 2:
 	 			Value = 3
 	 			break
 	 		case 3:
 	 			Value = 2
 	 			break
			default:
				let _ = 0
 		}
 		UpdateInnerValues()
 		UpdateOuterValues()
 		return self
 	}

 	@discardableResult
	public override func Move() -> CValueSequencer {
		if Value == 0 {
 	 		Value = 1
		}
 		else if Value == 3 {
			Value = 2
		}
 		UpdateInnerValues()
 		UpdateOuterValues()
 		return self
 	}

 	@discardableResult
	public override func Young() -> CValueSequencer {
		if Value == 0 {
 	 		Value = 2
		}
		else if Value == 3 {
			Value = 1
		}
 		UpdateInnerValues()
 		UpdateOuterValues()
 		return self
 	}

 	@discardableResult
	public override func Old() -> CValueSequencer {
		if Value == 2 {
 	 		Value = 0
		}
		else if Value == 1 {
			Value = 3
		}
 		UpdateInnerValues()
 		UpdateOuterValues()
 		return self
 	}

 	public override func UpdateInnerValues() {
 		switch (Value)
 		{
 	 		case 0:
				Bit(nIndex: 2).Value = 0
 	 			Bit(nIndex: 1).Value = 0
 	 			Bit(nIndex: 0).Value = 0
 	 			//break
 	 		case 1:
 	 			Bit(nIndex: 2).Value = 0
 	 			Bit(nIndex: 1).Value = 1
 	 			Bit(nIndex: 0).Value = 0
 	 			break
 	 		case 2:
 	 			Bit(nIndex: 2).Value = 1
 	 			Bit(nIndex: 1).Value = 0
 	 			Bit(nIndex: 0).Value = 1
 	 			break
 	 		case 3:
 	 			Bit(nIndex: 2).Value = 1
 	 			Bit(nIndex: 1).Value = 1
 	 			Bit(nIndex: 0).Value = 1
 	 			break
			default:
				let _ = 0
 		}
 		Bit(nIndex: 2).UpdateInnerValues()
 		Bit(nIndex: 1).UpdateInnerValues()
 		Bit(nIndex: 0).UpdateInnerValues()
 	}

 	public override func UpdateOuterValues() {
 		Value = Bit(nIndex: 0).Value + Bit(nIndex: 1).Value + Bit(nIndex: 2).Value
 		m_pvsParent?.UpdateOuterValues()
 	}

	public static func SetCurrentSequence(nSequence: Int) {
 		m_nCurrentSequence = nSequence
 	}

	public static func SetCurrentRatio(nRatio: Int) {
 		m_nCurrentRatio = nRatio
 	}

	public static func SetCurrentLabel(nLabel: Int) {
 		m_nCurrentLabel = nLabel
 	}

 	public override func GetLabel() -> String {
 		return "Sequences.strLineLabels[GetCurrentLabel(),Value]"
 	}

 	public override func GetMoving() -> Bool {
 		return (Value == 0) || (Value == 3)
 	}

	public override func GetCurrentSequence() -> Int { return CLineValueSequencer.m_nCurrentSequence }
	public override func GetCurrentRatio() -> Int { return CLineValueSequencer.m_nCurrentRatio }
	public override func GetCurrentLabel() -> Int { return CLineValueSequencer.m_nCurrentLabel }

	private static var m_nCurrentSequence: Int = 1 /////////// ?????????????????
 	private static var m_nCurrentRatio: Int = 0
 	private static var m_nCurrentLabel: Int = 0
}

File: value_sequencer_py/line_value_sequencer.py
from __future__ import annotations

from typing import List

from value_sequencer_py.bit_value_sequencer import CBitValueSequencer
from value_sequencer_py.value_sequencer import CValueSequencer
from value_sequencer_py.sequences import Sequences


class CLineValueSequencer(CValueSequencer):
    _nCurrentSequence = 0
    _nCurrentRatio = 0
    _nCurrentLabel = 0

    def __init__(self, value: int) -> None:
        super().__init__(3, 4, value)
        self._pvsInner = [CBitValueSequencer(0) for _ in range(3)]
        for bit in self._pvsInner:
            bit.set_parent(self)
        self._nSequences = Sequences.nLineSequences
        self._nRatios = Sequences.nLineRatios

    def bit(self, index: int) -> CBitValueSequencer:
        return self._pvsInner[index]  # type: ignore[return-value]

    def inverse(self) -> "CLineValueSequencer":
        mapping = {0: 1, 1: 0, 2: 3, 3: 2}
        self.value = mapping.get(self.value, self.value)
        self.update_inner_values()
        self.update_outer_values()
        return self

    def move(self) -> "CLineValueSequencer":
        if self.value == 0:
            self.value = 1
        elif self.value == 3:
            self.value = 2
        self.update_inner_values()
        self.update_outer_values()
        return self

    def young(self) -> "CLineValueSequencer":
        if self.value == 0:
            self.value = 2
        elif self.value == 3:
            self.value = 1
        self.update_inner_values()
        self.update_outer_values()
        return self

    def old(self) -> "CLineValueSequencer":
        if self.value == 2:
            self.value = 0
        elif self.value == 1:
            self.value = 3
        self.update_inner_values()
        self.update_outer_values()
        return self

    def update_inner_values(self) -> None:
        configurations: List[List[int]] = [
            [0, 0, 0],
            [0, 1, 0],
            [1, 0, 1],
            [1, 1, 1],
        ]
        bit_values = configurations[self.value]
        for index, bit in enumerate(reversed(self._pvsInner)):
            bit.value = bit_values[index]
            bit.update_inner_values()

    def update_outer_values(self) -> None:
        self.value = sum(bit.value for bit in self._pvsInner)
        if self._pvsParent:
            self._pvsParent.update_outer_values()

    @classmethod
    def set_current_sequence(cls, sequence: int) -> None:
        cls._nCurrentSequence = sequence

    @classmethod
    def set_current_ratio(cls, ratio: int) -> None:
        cls._nCurrentRatio = ratio

    @classmethod
    def set_current_label(cls, label: int) -> None:
        cls._nCurrentLabel = label

    def get_label(self) -> str:
        return Sequences.strLineLabels[self.get_current_label()][self.value]

    def get_moving(self) -> bool:
        return self.value in {0, 3}

    def get_current_sequence(self) -> int:
        return CLineValueSequencer._nCurrentSequence

    def get_current_ratio(self) -> int:
        return CLineValueSequencer._nCurrentRatio

    def get_current_label(self) -> int:
        return CLineValueSequencer._nCurrentLabel
